ui/types/api/qtree.d.ts

```diff
@@ -1,24 +1,28 @@
-// We can't set QTreeNode as default assignment for the generics param,
-// we use unknown and use conditional types to set it
-export interface QTreeNode<Node = unknown> {
-  label?: string;
-  icon?: string;
-  iconColor?: string;
-  img?: string;
-  avatar?: string;
-  children?: (Node extends unknown ? QTreeNode : Node)[];
-  disabled?: boolean;
-  expandable?: boolean;
-  selectable?: boolean;
-  handler?: (node: Node extends unknown ? QTreeNode : Node) => void;
-  tickable?: boolean;
-  noTick?: boolean;
-  tickStrategy?: "leaf" | "leaf-filtered" | "string" | "none";
-  lazy?: boolean;
-  header?: string;
-  body?: string;
-  [index: string]: any;
-}
+/**
+ * @template TExtra Object type to add extra properties for the node, overrides the existing ones as well
+ */
+export type QTreeNode<TExtra = unknown> = Omit<
+  {
+    label?: string;
+    icon?: string;
+    iconColor?: string;
+    img?: string;
+    avatar?: string;
+    children?: QTreeNode<TExtra>[];
```

Comment:
This implementation change will break the use case where you have children which have different properties than the parent, or am I incorrect?
Maybe we can add 2 generics, one for the children Node type (which defaults to the same type of the parent) and one for extra properties on the current node

Reply:
It is possible to override the `children` type like this:
```ts
type ChildNode = QTreeNode<{ foo: string }>
type ParentNode = QTreeNode<{ bar: number; children?: ChildNode[] }>
```
Sure, we can add another generic type parameter, but I think the existing solution is simpler and more flexible. You can make the children required this way, or even use a tuple type to have a known child sequence, etc.
```ts
type ChildNode = QTreeNode<{ foo: string }>
// required children
type ParentNode = QTreeNode<{ bar: number; children: ChildNode[] }>
```
```ts
type ChildNode1 = QTreeNode<{ foo: string }>
type ChildNode2 = QTreeNode<{ baz: string }>
// required children with specific children count/order/types
type ParentNode = QTreeNode<{ bar: number; children: [ChildNode1, ChildNode2] }>
```

Reply:
Here is an example for representing the basics of a file system:
```ts
type File = {
  path: string;
  size: number;
  lastModified: number;
}
type FileNode = QTreeNode<File & { type: "file", children?: never }>;
type DirectoryNode = QTreeNode<File & { type: "directory", children?: (FileNode | DirectoryNode)[] }>;

const directory: DirectoryNode = {
  type: "directory",
  path: "/",
  size: 0,
  lastModified: 0,
  // allows both file and directory as children
  children: [
    {
      type: "file",
      path: "/foo.txt",
      size: 100,
      lastModified: 1000,
      // does not allow children
    },
    {
      type: "directory",
      path: "/bar",
      size: 0,
      lastModified: 0,
      // empty folder - doesn't have children
    }
  ]
}
```

Reply:
That's actually super smart and flexible, but not so clear
Can we add some documentation about this? Or at least add these examples in the original issue

Documentation aside, LGTM at this point

Reply:
Another advantage is allowing children to type when using a different children key. The `children` property can be different(e.g. `nodes`), which can be modified through `children-key`. So, this solution also allows easily adding the children types with different children property names as well.

Reply:
Added some documentation with examples in the form of JSDocs